### PR TITLE
8333258: C2: high memory usage in PhaseCFG::insert_anti_dependences()

### PR DESCRIPTION
**Pull request by rwestrel**
In a debug build, `PhaseCFG::insert_anti_dependences()` is called
twice for a single node: once for actual processing, once for
verification.

In TestAntiDependenciesHighMemUsage, the test has a `Region` that
merges 337 incoming path. It also has one `Phi` per memory slice that
are stored to: 1000 `Phi` nodes. Each `Phi` node has 337 inputs that
are identical except for one. The common input is the memory state on
method entry. The test has 60 `Load` that needs to be processed for
anti dependences. All `Load` share the same memory input: the memory
state on method entry. For each `Load`, all `Phi` nodes are pushed 336
times on the work lists for anti dependence processing because all of
them appear multiple times as uses of each `Load`s memory state: `Phi`s
are pushed 336 000 on 2 work lists. Memory is not reclaimed on exit
from `PhaseCFG::insert_anti_dependences()` so memory usage grows as
`Load` nodes are processed:

336000 * 2 work lists * 60 loads * 8 bytes pointer = 322 MB. 

The fix I propose for this is to not push `Phi` nodes more than once
when they have the same inputs multiple times.

In TestAntiDependenciesHighMemUsage2, the test has 4000 loads. For
each of them, when processed for anti dependences, all 4000 loads are
pushed on the work lists because they share the same memory
input. Then when they are popped from the work list, they are
discarded because only stores are of interest:

4000 loads processed * 4000 loads pushed * 2 work lists * 8 bytes pointer = 256 MB. 

The fix I propose for this is to test before pushing on the work list
whether a node is a store or not.

Finally, I propose adding a `ResourceMark` so memory doesn't
accumulate over calls to `PhaseCFG::insert_anti_dependences()`.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8333258](https://bugs.openjdk.org/browse/JDK-8333258): C2: high memory usage in PhaseCFG::insert_anti_dependences() (**Bug** - P4)

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)
 * [Emanuel Peter](https://openjdk.org/census#epeter) (@eme64 - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19791/head:pull/19791` \
`$ git checkout pull/19791`

Update a local copy of the PR: \
`$ git checkout pull/19791` \
`$ git pull https://git.openjdk.org/jdk.git pull/19791/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19791`

View PR using the GUI difftool: \
`$ git pr show -t 19791`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19791.diff">https://git.openjdk.org/jdk/pull/19791.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19791#issuecomment-2178666708)

### CODE REVIEW

**Review by vnkozlov — CHANGES_REQUESTED** [review]
We should use `VectorSet` for something like that.
`worklist_visited` is definitely should be `VectorSet`.
Instead of `already_enqueued` we should add `VectorSet` for `mem` to check if it was processed before. This assumes that this loop don't modify graph. Unfortunately we have one modification `store->add_prec(load);` which we should do something about. I may miss other ocde which modify graph.

**Comment by rwestrel** [comment]
> We should use `VectorSet` for something like that. `worklist_visited` is definitely should be `VectorSet`. Instead of `already_enqueued` we should add `VectorSet` for `mem` to check if it was processed before. This assumes that this loop don't modify graph. Unfortunately we have one modification `store->add_prec(load);` which we should do something about. I may miss other ocde which modify graph.

I don't think `VectorSet` works here because It seems there are cases where we want to allow a Phi to be enqueued more than once. There's this comment:
```
// (Allow phis to be repeated; they can merge two relevant states.)
```
and in this loop when processing Phis:
```
      for (uint j = PhiNode::Input, jmax = store->req(); j < jmax; j++) {
        if (store->in(j) == mem) {   // Found matching input?
```
couldn't we reach the Phi from another mem input?
What the patch I propose does is that it prevents a Phi from being enqueued multiple times when it's reached from a particular mem input.

**Comment by eme64** [comment]
Why does `VectorSet` not work for `worklist_visited.size`? I think there are not any duplicates in this list. We only once add the first element:
`worklist_visited.push(initial_mem);`
And we add elements like this:
```
          // Be sure we don't get into combinatorial problems.
          // (Allow phis to be repeated; they can merge two relevant states.)
          uint j = worklist_visited.size();
          for (; j > 0; j--) {
            if (worklist_visited.at(j-1) == store)  break;
          }
          if (j > 0)  continue; // already on work list; do not repeat
          worklist_visited.push(store);
```
In my eye, this makes `worklist_visited` unique, and so `VectorSet` would be preferrable.

The comment you quoted says that there can be multiple `phi`, but it does not say anything about `store`, right?

**Comment by rwestrel** [comment]
VectorSet would work for worklist_visited but it's not the thing that I'm trying to fix here. @vnkozlov's comment is not only about worklist_visited.

Also if there are few uses, the existing code would only have to go over a few `worklist_visited` elements while using a `VectorSet` would possibly require zeroing a small array. Is it that clear cut that a `VectorSet` is always better?

**Comment by eme64** [comment]
Never mind, I removed my comment. I need to understand the whole algorithm better first.
I guess it could be a tradeoff: a `Node_List` is here "sparse", it only requires as much space as you push elements. But a `VectorSet` is more compact if you add a lot of elements, plus access time is faster.

Of course that could be investigated separately.

**Comment by eme64** [comment]
First a question:
Why does your `already_enqueued` exit early if `worklist_mem.at(j-1) != mem`? Are we sure that there will not be a `mem, store` pair at a lower index `i`?

An idea: this really looks like we are storing a `Pair<mem, store>`. Could we not use a set of some kind, maybe a `ResourceHashtable`? That would allow a possibly faster lookup. Ah, or maybe you can combine the `idx` of `mem` and `store` to a unique combined idx, and use that for a `VectorSet`?

I just think that a hand-crafted `already_enqueued` is not really super nice, especially if it has this early-exit mechanism that is not really predictable from the name.

**Comment by eme64** [comment]
@rwestrel I'm sure that most of my questions / babbel above is because I don't understand the algorithm yet. Feel free to summarize it. But I think I understand the general idea: we are trying to limit duplicate edge visits. Not sure if in all cases or only in special cases and why?

I think you could add a nice explanatory comment, about what edges we need to traverse, and why and where we limit the duplication.

**Comment by rwestrel** [comment]
> First a question: Why does your `already_enqueued` exit early if `worklist_mem.at(j-1) != mem`? Are we sure that there will not be a `mem, store` pair at a lower index `i`?

I ran testing with some verification code that checked that and it never failed. But in any case, that doesn't really matter. The logic I've added tries to prevent a pathological case from happening. There's no correctness issue involved. So if once in a while we re-enqueue a `mem, store`, that doesn't make much of a difference.

> An idea: this really looks like we are storing a `Pair<mem, store>`. Could we not use a set of some kind, maybe a `ResourceHashtable`? That would allow a possibly faster lookup. Ah, or maybe you can combine the `idx` of `mem` and `store` to a unique combined idx, and use that for a `VectorSet`?

Yes, it is a `Pair<mem, store>`. What you're describing seems like we're trying to solve a problem that we don't know exist. Shouldn't we stick to simplest possible implementation unless it's shown to not perform well? I just ran ctw on java.base and looked at the CITime output. I don't see a difference with or without this patch.

File: src/hotspot/share/opto/gcm.cpp [hunk 1]
```diff
@@ -686,7 +710,6 @@ Block* PhaseCFG::insert_anti_dependences(Block* LCA, Node* load, bool verify) {
     }
 
     if (op == Op_MachProj || op == Op_Catch)   continue;
-    if (store->needs_anti_dependence_check())  continue;  // not really a store
```
[the remark this answers is not in the document]
Comment (rwestrel) [reply src/hotspot/share/opto/gcm.cpp:689]: I didn't remove it, I moved it. It's what this:

```
The fix I propose for this is to test before pushing on the work list
whether a node is a store or not.
```

refers to. What we do today is that we push all use and then filter out what's not of interest. What I propose is to filter out what's not useful before it's pushed so the queue doesn't grow large with nodes that are going to be discarded anyway.

[the remark this answers is not in the document]
Comment (rwestrel) [reply src/hotspot/share/opto/gcm.cpp:689]: That's what I did but the assert is where the loop starts not where the line was removed.

**Comment by eme64** [comment]
Fair enough. Maybe the code is ok as it is. Though the naming is quite bad, but that is not really your fault.
The more code we add the harder it will be to read.
So I would appreciate at least some more comments that help the reader.

File: src/hotspot/share/opto/gcm.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (eme64) [reply src/hotspot/share/opto/gcm.cpp:689]: Hmm, I did see that. But that is quite far away, and who knows maybe someone in the future breaks that assumption somewhere inside the loop. Therefore, it would be nice and easy to restate that assumption.

File: src/hotspot/share/opto/gcm.cpp [hunk 2]
```diff
@@ -562,6 +562,35 @@ bool PhaseCFG::unrelated_load_in_store_null_block(Node* store, Node* load) {
...
+  // (mem,store) where mem is pushed to worklist_mem and store is
+  // pushed to worklist_store and store is a use of mem. Anytime a mem
+  // is pushed/popped to/from worklist_mem, a store has to be
+  // pushed/popped to/from worklist_store.
+  // If there are any use of mem already enqueued, they were enqueued
+  // last (all use of mem are processed in one go).
+  for (; j > 0; j--) {
+    if (worklist_mem.at(j-1) != mem) {
+      // We're done with the uses of mem
+      return false;
```
Comment (vnkozlov) [inline src/hotspot/share/opto/gcm.cpp:585]: Should this be `continue;` instead to make sure we scan all memory nodes?

Comment (eme64) [reply src/hotspot/share/opto/gcm.cpp:585]: Seems my old comments have disappeared, and with them Roland's answers.

Roland said duplicates are ok. But if you @vnkozlov are asking, then @rwestrel needs to make a comment about why we can early exit here ;)

Comment (eme64) [reply src/hotspot/share/opto/gcm.cpp:585]: Ah, it the comment is still here:
https://github.com/openjdk/jdk/pull/19791#issuecomment-2180754800

Comment (vnkozlov) [reply src/hotspot/share/opto/gcm.cpp:585]: I think the last statement in comment is not correct.
You can push several `(mem, store)` with the same memory. Then you pop only one top pair and start processing it which leads to different memory in `(mem, store)` pairs pushed on worklist. After processing some of pairs you may see original memory node again.

Comment (vnkozlov) [reply src/hotspot/share/opto/gcm.cpp:585]: Thank you, @eme64.  The last statement have to be adjusted based on Roland's answer you pointed.

Comment (rwestrel) [reply src/hotspot/share/opto/gcm.cpp:585]: We start by pushing:
```
(null, initial_mem)
```
then we pop it and push all uses of `initial_mem` before we pop anything:
```
(initial_mem, store_1)
(initial_mem, store_2)
(initial_mem, store_3)
...
(initial_mem, store_n-1)
(initial_mem, store_n)
```
Then we pop a pair:
```
(initial_mem, store_n)
```
And we push uses of `store_n` only if it is a `MergeMem`
```
(initial_mem, store1)
(initial_mem, store2)
(initial_mem, store3)
...
(initial_mem, storen-1)
(store_n, store_n_1)
...
(store_n, store_n_n)
```
And we keep going.
What you're saying is we can end up with:
```
(mem1, ...)
(mem2, ...)
(mem1, ...)
```
For that `mem1` and `mem2` have to be a `MergeMem`.  So There has to be a cycle from `mem1` to `mem1` through `mem2` all of which are `MergeMem`. That can't happen right? Or am I missing something?

Comment (vnkozlov) [reply src/hotspot/share/opto/gcm.cpp:585]: > And we push uses of store_n only if it is a MergeMem
You are right, I missed `op == Op_MergeMem ` check.

Yes, I thought about loop since we check for Phi node. But we have `worklist_visited` check for `MergeMem` which will not allow to push the same `mem1` on work list. It look like you are right - there should be only sequential elements in work list for the same memory.
I withdraw my comment.

**Review by vnkozlov — APPROVED** [review]

**Review by eme64 — COMMENTED** [review]
@rwestrel thanks for the updates! Sorry it took me a while to come back to this.

I think most likely your fix is correct. But I have a few complaints / suggestions, some of which I have already mentioned but you did not yet respond to - maybe you disagree with them. Sorry, I'll be a little annoying now :grimacing: 

- The naming of `store` in this code is really bad. Rather than `mem` and `store`, it should be `use` and `def`. Because apparently a `store` can also be a `Phi` - that is not very intuitive.
- This bad naming was there before your fix, but you are leaning into it with code like this, and this just makes the code even more confusing than it already is:
```
        } else if (store->is_Phi()) {
          // A Phi could have the same mem as input multiple times. If that's the case, we don't need to enqueue it
          // more than once.
          if (already_enqueued(worklist_mem, worklist_store, mem, store->as_Phi())) {
            continue;
          }
```
And:
```
assert(!store->needs_anti_dependence_check(), "only stores");
```
But it turns out that also a `MergeMem` has no anti-dependence-check, as the code suggests. So your assert text should be more accurate.

- Existing comments are half-accurate it seems:
```
  // Therefore, the branches visited by the worklist are of this form:
  //    initial_mem -> (MergeMem ->)* store
```
There are not just `MergeMem` on those branches, but also `Phi`s, right? While we are working on this code, we might as well fix that.

- For clarity, I would rename the inputs of `already_enqueued`:
```
static bool already_enqueued(const Node_List& worklist_mem, const Node_List& worklist_store, Node* mem, PhiNode* phi) {
->
static bool already_enqueued(const Node_List& worklist_mem, const Node_List& worklist_store, Node* def_mem, PhiNode* use_phi) {
```

- Another refactoring suggestion: Rather than having lengthy comments about saying that `worklist_mem/worklist_store` is a pair of lists, I would make it into a list of pairs. This could be called `edge_worklist` or `use_def_worklist`. Maybe there are even memory benefits because of better co-location.

- You removed the last:
```
if (store->needs_anti_dependence_check())  continue;  // not really a store
```
That is fine, but I would replace it with an assert, this helps the reader to know that after the traversal loop we actually have some node with that characteristic.

- The argument you are making in `already_enqueued` (that all of `mem`'s uses are processed in one go) is quite implicit. How much runtime would we actually lose if we scanned the whole worklist instead? I'm just wondering if this is not a premature optimization, that then requires an argument that is not clear to the reader (it was not immediately clear to me and also not to Vladimir K).

**Comment by rwestrel** [comment]
>     * The naming of `store` in this code is really bad. Rather than `mem` and `store`, it should be `use` and `def`. Because apparently a `store` can also be a `Phi` - that is not very intuitive.

I don't quite agree. I agree `store` is not great but people reading the code likely think about a `StoreNode` which is not always the case. But `use` is not better because it's too vague. It's not any kind of `use`. It's a `use` that also changing memory state (and in that respect it acts as a `store`).

>     * You removed the last:
> 
> 
> ```
> if (store->needs_anti_dependence_check())  continue;  // not really a store
> ```
> 
> That is fine, but I would replace it with an assert, this helps the reader to know that after the traversal loop we actually have some node with that characteristic.

The assert is on loop entry. There's a if between the new assert and the condition that was removed but the if block ends with a continue. So the assert is guaranteed to be executed every time the removed was executed.

>     * The argument you are making in `already_enqueued` (that all of `mem`'s uses are processed in one go) is quite implicit. How much runtime would we actually lose if we scanned the whole worklist instead? I'm just wondering if this is not a premature optimization, that then requires an argument that is not clear to the reader (it was not immediately clear to me and also not to Vladimir K).

I could add debug code that goes over the rest of the list and checks it doesn't find that pair enqueued elsewhere.

**Comment by rwestrel** [comment]
The new commit refactors the work queues so there's no longer a pair of queues but single queue of pairs.

**Comment by rwestrel** [comment]
@eme64 what do you think of the refactoring & answers to your questions?

**Comment by eme64** [comment]
@rwestrel I'm sorry I didn't get back to this earlier.

It looks better, I like the `MemStoreQueue` and assert at the end of `already_enqueued`.

> > ```
> > * The naming of `store` in this code is really bad. Rather than `mem` and `store`, it should be `use` and `def`. Because apparently a `store` can also be a `Phi` - that is not very intuitive.
> > ```
> 
> 
> I don't quite agree. I agree `store` is not great but people reading the code likely think about a `StoreNode` which is not always the case. But `use` is not better because it's too vague. It's not any kind of `use`. It's a `use` that also changing memory state (and in that respect it acts as a `store`).

Hmm. Personally I'd rather have something more vague than misleading. Maybe it could be `store -> use_mem_state` and `mem -> def_mem_state`? I guess that is then a bit wordy. Up to you.

```
sounds strange:
store->is_MergeMem()
store->is_Phi()

sounds more intuitive:
use_mem_state->is_MergeMem()
use_mem_state->is_Phi()
```

> The assert is on loop entry. There's a if between the new assert and the condition that was removed but the if block ends with a continue. So the assert is guaranteed to be executed every time the removed was executed.

I understand, and agree on a technical level. Someone in the future may break things, and that is why I would prefer the assert to be there. But up to you.

**Comment by eme64** [comment]
But I would like you to fix the comments here:
```
  // The relevant stores "nearby" the load consist of a tree rooted
  // at initial_mem, with internal nodes of type MergeMem.
  // Therefore, the branches visited by the worklist are of this form:
  //    initial_mem -> (MergeMem ->)* store
  // The anti-dependence constraints apply only to the fringe of this tree.
```
There are not just `MergeMem` but also `Phi` nodes.

**Comment by rwestrel** [comment]
> Hmm. Personally I'd rather have something more vague than misleading. Maybe it could be `store -> use_mem_state` and `mem -> def_mem_state`? I guess that is then a bit wordy. Up to you.

I went with that. Thanks for the suggestion.

> > The assert is on loop entry. There's a if between the new assert and the condition that was removed but the if block ends with a continue. So the assert is guaranteed to be executed every time the removed was executed.
> 
> I understand, and agree on a technical level. Someone in the future may break things, and that is why I would prefer the assert to be there. But up to you.

The assert on loop entry is executed more often (and on every element of the queue) than if it was where you recommend it to be which is why I'd like to leave it there.

**Comment by rwestrel** [comment]
> But I would like you to fix the comments here:
> 
> ```
>   // The relevant stores "nearby" the load consist of a tree rooted
>   // at initial_mem, with internal nodes of type MergeMem.
>   // Therefore, the branches visited by the worklist are of this form:
>   //    initial_mem -> (MergeMem ->)* store
>   // The anti-dependence constraints apply only to the fringe of this tree.
> ```
> 
> There are not just `MergeMem` but also `Phi` nodes.

I updated the comment. Does that look better to you?

**Comment by rwestrel** [comment]
@eme64 can you have a look at the updated change?

**Comment by rwestrel** [comment]
@eme64 could you take another look at this one?

File: src/hotspot/share/opto/gcm.cpp [hunk 3]
```diff
@@ -644,60 +744,55 @@ Block* PhaseCFG::insert_anti_dependences(Block* LCA, Node* load, bool verify) {
...
     // Examine a nearby store to see if it might interfere with our load.
-    Node* mem   = worklist_mem.pop();
-    Node* store = worklist_store.pop();
-    uint op = store->Opcode();
+    Node* def_mem_state = worklist_def_use_mem_states.top_def();
+    Node* use_mem_state = worklist_def_use_mem_states.top_use();
+    worklist_def_use_mem_states.pop();
+
+    uint op = use_mem_state->Opcode();
+    assert(!use_mem_state->needs_anti_dependence_check(), "only stores");
```
Comment (eme64) [inline src/hotspot/share/opto/gcm.cpp:761]: The comment in the assert is misleading. It can also be MergeMem, Phi, etc

File: src/hotspot/share/opto/gcm.cpp [hunk 4]
```diff
@@ -644,60 +744,55 @@ Block* PhaseCFG::insert_anti_dependences(Block* LCA, Node* load, bool verify) {
   // The relevant stores "nearby" the load consist of a tree rooted
   // at initial_mem, with internal nodes of type MergeMem.
   // Therefore, the branches visited by the worklist are of this form:
-  //    initial_mem -> (MergeMem ->)* store
+  //    initial_mem -> (MergeMem ->)* Memory state modifying node
+  // Memory state modifying nodes include Store and Phi nodes and any node for which needs_anti_dependence_check()
+  // returns true.
```
Comment (eme64) [inline src/hotspot/share/opto/gcm.cpp:749]: Basically this is the idea:
`needs_anti_dependence_check() == true` for Loads
`needs_anti_dependence_check() == false` for MergeMem, Phi, Store

So your `returns true` is not correct, right? Should it not be `returns false`?

File: src/hotspot/share/opto/gcm.cpp (hunk 3 above)
Comment (eme64) [reply src/hotspot/share/opto/gcm.cpp:761]: Maybe say `only MergeMem and state modifying nodes`. Or just `no loads`?

File: src/hotspot/share/opto/gcm.cpp [hunk 5]
```diff
@@ -644,60 +744,55 @@ Block* PhaseCFG::insert_anti_dependences(Block* LCA, Node* load, bool verify) {
...
-          // (Allow phis to be repeated; they can merge two relevant states.)
-          uint j = worklist_visited.size();
-          for (; j > 0; j--) {
-            if (worklist_visited.at(j-1) == store)  break;
-          }
-          if (j > 0)  continue; // already on work list; do not repeat
-          worklist_visited.push(store);
+      for (DUIterator_Fast imax, i = def_mem_state->fast_outs(imax); i < imax; i++) {
+        use_mem_state = def_mem_state->fast_out(i);
+        // If this is not a store, load can't be anti dependent on this node
```
Comment (eme64) [inline src/hotspot/share/opto/gcm.cpp:775]: This comment is not helping me, and I think it is also not precise...

Instead of `store`, you should be talking about `MergeMem`, `Phi`, `Store` and any other state modifying node.

Suggestion: remove comment, and after the check, just before the `continue`, add this comment:
```
use_mem_state is also a kind of load (i.e. needs_anti_dependence_check), and it is not any state modifying node, store, Phi or MergeMem. Hence, load is not anti dependent on this node.
```

File: src/hotspot/share/opto/gcm.cpp [hunk 6]
```diff
@@ -562,6 +562,107 @@ bool PhaseCFG::unrelated_load_in_store_null_block(Node* store, Node* load) {
...
+
+  void push(Node* def_mem_state, Node* use_mem_state) {
+    if (use_mem_state->is_MergeMem()) {
+      // Be sure we don't get into combinatorial problems.
+      // (Allow phis to be repeated; they can merge two relevant states.)
+      uint j = _worklist_visited.size();
+      for (; j > 0; j--) {
+        if (_worklist_visited.at(j-1) == use_mem_state)  return; // already on work list; do not repeat
+      }
+      _worklist_visited.push(use_mem_state);
```
Comment (eme64) [inline src/hotspot/share/opto/gcm.cpp:637]: This is fine, but it might be nicer if we had such a `push_if_missing`. `GrowableArray` has that functionality. Maybe you should just use a `GrowableArray` then anyway, right? It would be less code.

**Review by eme64 — CHANGES_REQUESTED** [review]
@rwestrel sorry for the long delay, I was away for over 3 weeks. Looks much better now, I still have a few comments and suggestions.

**Comment by rwestrel** [comment]
@eme64 can you have another look. New commit should address your comments.

**Review by eme64 — APPROVED** [review]
Thanks for all the updates @rwestrel . LGTM :)

**Comment by rwestrel** [comment]
@eme64 thanks for the review
@vnkozlov does this still look good to you?

**Review by vnkozlov — APPROVED** [review]
Latest changes look good to me.

**Comment by rwestrel** [comment]
@vnkozlov thanks for the re-review!
